utensor_cgen/legalizer/tflite.py

```diff
@@ -0,0 +1,138 @@
+import re
+from copy import deepcopy
+from functools import reduce
+
+from utensor_cgen.ir.base import OperationInfo, TensorInfo
+from utensor_cgen.utils import topologic_order_graph
+
+from .api import Legalizer
+from .base import LegalizerBase
+
+
+@Legalizer.register
+class TFLiteLegalizer(LegalizerBase):
+  TARGET = "tflite"
+
+  def legalize_ops(self, ugraph):
+    _GraphRewrite.apply(ugraph)
+    _OpTypeRename.apply(ugraph)
+    return ugraph
+  
+  def legalize_dtype(self, ugraph):
+    return ugraph
+
+
+class _OpTypeRename(object):
```

Comment:
These op names should not auto rename to quantized forms. This should be handled by the graph lowering stage.
`FullyConnected` -> `FullyConnectedOperator` as the generic form. Then the later stages can check quantization based on the tensors + quantization attributes.

Reply:
I agree.
I'll see how to fix it.

Reply:
But I don't think we should force the generic form to be unquantized.
Many models are given quantized and there is no guarantee we can convert quantized model back to a generic unquantized form.
I believe there is use case where the generic form is a quantized graph.

Reply:
I'm not suggesting forcing the generic form to be unquantized, I'm suggesting we leave the name to represent the op type and inject any attributes that can be used by the transform tools. For example:

```
Conv2D(input: float, filter: quantized int8, bias: int32)
```
Is probably not legal from a uTensor perspective so an error should be thrown since not all inputs are quantized.

But the following are both legal from a IR perspective, since they both represent the same functional operation even though their lowered targets are slightly different:
```
Conv2D(input: float, filter: float, bias: float); // Gets lowered to a Reference operator
Conv2D(input: quantized int8, filter: quantized int8, bias: int32) // Same op, but is_quantized is set to true so get's lowered to best form that accepts quantized inputs
```

Really the only stage between the legalizer and backend lowering is the Transformer pipeline, so as long as we capture intent of the op then the transforms can operate correctly. This way, the transforms can impose any additional constraints like all operators are quantized, or none, or a mix, without having to maintain some unique name for each type of op configuration.

Reply:
I'm pretty sure I fixed the legalizer and lower.
Now the legalizer will not assuming a quantized graph and lower will change the op type once it find out the given graph is quantized.
BTW, one heuristic here is that I assume the graph is quantized if there is any tensor with quantization parameters.
I think this assumption is fair enough in most cases.

Reply:
Update: the graph lower will quantize the graph op by op.